### DISCUSSION TITLE
Where is "com.xilinx.rapidwright.bitstream" ?

### DISCUSSION THREAD

**Discussion by tysung** [Xilinx/RapidWright#316] [Q&A] [answered]
Checking the release note, I noticed that there are "Bitstream" class released already.
But with the updated repo in master branch, I couldn't locate it.
Can you tell me where can I find it in the RapidWright repo?

============= RapidWright 2020.2.0-beta released on 2021-01-15 ================
Notes:
 * Adds a set of bitstream manipulation APIs
 * Adds part name normalization (see pull request #120)
 * Bug fix for migrateCellAndSubCells() (see pull request #116)
 * Fixes an NPE on getSitePinFromLogicalPin()
 * Fixes support for BUFCE_ROW on Net.{get,set}BufferDelay()
 - API Additions:
    - com.xilinx.rapidwright.bitstream.{BitLocation, Bitstream, BitstreamHeader, Block,
                                        BlockSubType, BlockType, CMDCode, ConfigArray, ConfigRow,
                                        FAR, Frame, IDCode, OpCode, Packet, PacketType,
                                        RegisterType} # Please see Javadocs for full details.

Thanks!
Ting-Yuan

**Comment by clavin-xlnx** [comment]
Hi Ting-Yuan, RapidWright does have some Bitstream parsing and manipulation capabilities represented as APIs.  Unfortunately, the source code for these APIs is not available at the present time.  We do have a few examples and Javadocs that describe how these can be used:

- Description and a few code examples: http://www.rapidwright.io/docs/Bitstream_Manipulation.html
- Javadoc for Bitstream package: http://www.rapidwright.io/javadoc/com/xilinx/rapidwright/bitstream/package-summary.html

Let us know if you have further questions.

Chris

**Reply by yu-zou** [reply]
Hi Chris,

If I understand correctly, bitstream manipulation implementations are not done, only APIs are defined.
Does that mean bitstream manipulation capability is currently not ready to use?

If that's true, is there any beta version or dev branch so that we can take a preliminary try?

Yu

**Reply by clavin-xlnx** [reply]
Hi Yu,

Limited functionality is available such as updating user state bits (LUTRAM and BRAM contents, for example).  Other than the APIs and functionality described in the references above, there has not been much new added lately.  One notable exception to this is some experimental work on bitstream relocation: https://github.com/Xilinx/RapidWright/blob/master/src/com/xilinx/rapidwright/util/RelocateBitstreamByRow.java.  This is still under development, but this code has been made available.  

Chris

**Reply by yu-zou** [reply]
Thanks. Glad to hear that bitstream manipulation is ready to use, even though currently the functionality is a little bit limited.
One more question. I'm not very familiar with Java, so this might sound stupid.

I see example bitstream manipulation code is using "import com.xilinx.rapidwright.bitstream.*;", but under com/xilinx/rapidwright folder, I could not find any folder named as bitstream. But surprisingly, the code is still able to be compiled.
I'm curious where is the bitstream class defined in the codebase? I could not find where the class is defined.

Yu

**Reply by clavin-xlnx** [reply, marked as answer]
Some functionality in RapidWright is distributed as compiled byte code (`*.class` files).  This portion of RapidWright is not open source, but a compiled library (`*.jar` file) that is distributed as a companion to the open source portion.  In this case, the bitstream package is currently not in the open source portion but in the compiled library portion.
